torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -161,8 +161,22 @@ c10::intrusive_ptr<RRef> RRefContext::getOrCreateRRef(
   auto& rrefId = rrefForkData.rrefId_;
   auto& forkId = rrefForkData.forkId_;
   if (ownerId == getWorkerId()) {
+    // We have found the rref through the rrefId
     auto ownerRRef = getOwnerRRef(rrefId);
-    TORCH_INTERNAL_ASSERT(ownerRRef->type() == type);
+    // Now double check if the two types are matched
+    //
+    // Why we are special casing the check for tensor type here?
+    // this is because tensor types might get specialized on tensors when
+    // we pass inputs to the function, i.e. TensorType can filled with
+    // specific shape info, requires_grad info, etc. so the OwerRRef we
+    // found might already have those infos, but the `type` we passed in
+    // here is a plain TensorType, they are not equal relationship. In RPC
+    // we don't care the difference as we ser/de with just the plain TensorType.
+    if(type == TensorType::get()) {
+      TORCH_INTERNAL_ASSERT(ownerRRef->type()->isSubtypeOf(TensorType::get()));
+    } else {
+      TORCH_INTERNAL_ASSERT(ownerRRef->type() == type);
+    }
     return ownerRRef;
   } else {
     return createUserRRef(ownerId, rrefId, forkId, type);
```

Comment:
will it impact createUserRRef as well when type does not match exactly?

Reply:
I think no, this will only happen when there's already a RRef in the RRefContext and we compare the type between the existing one and the pass we passed in, we are always creating UserRRef instead of retrieving existing UserRRefs, so I think we are fine for this case.